### ISSUE TITLE
Adversary loss function

### ISSUE THREAD

**Issue by YueErro** [openai/baselines#453]
I wonder what's the difference between GAN and GAIL regarding loss function.
In many documents, i saw it is computed like this:
```diff
-    generator_loss = tf.nn.sigmoid_cross_entropy_with_logits(logits=generator_logits, labels=tf.zeros_like(generator_logits))
+    generator_loss = tf.nn.sigmoid_cross_entropy_with_logits(logits=generator_logits, labels=tf.ones_like(generator_logits))
     generator_loss = tf.reduce_mean(generator_loss)

-    expert_loss = tf.nn.sigmoid_cross_entropy_with_logits(logits=expert_logits, labels=tf.ones_like(expert_logits))
+    d_loss_real=tf.nn.sigmoid_cross_entropy_with_logits(logits=expert_logits, labels=tf.ones_like(expert_logits))
+    d_loss_fake=tf.nn.sigmoid_cross_entropy_with_logits(logits=generator_logits, labels=tf.zeros_like(generator_logits))
+    expert_loss = d_loss_real + d_loss_fake
     expert_loss = tf.reduce_mean(expert_loss)
```
Example: [Alternative Loss Formulation ](https://wiseodd.github.io/techblog/2016/09/17/gan-tensorflow/)section.

As they say in this document, the Generator wants to maximize the probability of fake data (that's why it's set to ones and not to zeros) and the Discriminator, on the one hand, it wants to maximize the probability of real data (d_loss_real set to ones), and on the other, it wants to minimize the probability of fake data (d_loss_fake set to zeros).

Are the changes correct?
I don't understand why in baselines the generator tries to minimize the fake data.

Thank you in advance.

**Comment by andrewliao11** [comment]
Hi Yue, 

Your understanding of adversarial loss is right, while in the adversary.py [L38~L49](https://github.com/openai/baselines/blob/master/baselines/gail/adversary.py#L38-L49), it's actually building the loss for the discriminator. The objective of the discriminator is to minimize the probability of generator (push the mass to zero) and maximize the probability of expert (push the mass to one).

To check the *true* objective of the generator, you should refer to the [reward signals](https://github.com/openai/baselines/blob/master/baselines/gail/adversary.py#L51). 

The variable naming might be a little bit misleading, where the `generator loss` is not the objective of the generator (policy).

**Comment by YueErro** [comment]
Thank you andrewliao, i close this issue.

**Comment by zhihaocheng** [comment]
@andrewliao11 ，hello Andrew, I refered to the reward signals. I am a little confused because according to the GAIL paper, the reward should be log(D), but the implementation is log(1-D). Does this matter a lot for GAIL in terms of reward form?
